### DISCUSSION TITLE
pyright fails to infer boundedness of conditionally bound variable for a statically known condition?

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#4866] [Q&A] [answered]
I have some code similar to the following:

```python
x = True

if x:
    y = "blah"

while x:
    y += "blah"  # error: "y" is possibly unbound (reportUnboundVariable)
```

Shouldn't it be known that `y` will be bound in the `while` loop given that the type of `x` is known to be `Literal[True]`? Could this also be extended to `bool` if conditions are identical as above?

pylint doesn't throw any errors in the above example, but it does when `x` is `Literal[False]`.

```python
x = False

if x:
    y = "blah"

while x:
    y += "blah"  # E0602: Undefined variable ‘y’ (undefined-variable)
```

`mypy --strict` doesn't complain in either case.

**Comment by erictraut** [comment, marked as answer]
Pyright doesn't track conditional type dependencies between variables like `x` and `y`. To do so, it would need to somehow determine that "the type of `y` is `str` if `x` is `Literal[True]` (and remains such) but is `Unbound` if `x` is `Literal[False]`. I'm not aware of any type analyzer in any language that does this. It's not computationally feasible.

I'm not sure how mypy's method for tracking unbound variables is implemented. It's relatively new, and there are known holes in it still, so I don't know whether the fact that it doesn't emit an error in this case is intentional or just a hole in its current implementation.

In any case, pyright is working as designed in this case.
